pybossa/sched.py

```diff
@@ -267,16 +268,42 @@ def template_get_locked_task(project_id, user_id=None, user_ip=None,
         sql = query_factory(project_id, user_id=user_id, limit=limit,
                             rand_within_priority=rand_within_priority,
                             task_type=task_type)
+        db_max_records = 100000
+        limit = db_max_records if filter_user_prefs else user_count + 5
         rows = session.execute(sql, dict(project_id=project_id,
                                          user_id=user_id,
                                          assign_user=assign_user,
-                                         limit=user_count + 5))
+                                         limit=limit))
+
+        if filter_user_prefs:
+            # calculate task preference score
+            user_profile = json.loads(cached_users.get_user_profile_metadata(user_id))
+            task_rank_info = []
+            for task_id, taskcount, n_answers, calibration, meta, timeout in rows:
+                score = 0
+                if meta:
+                    pref = json.loads(meta).get('preference', {})
+                    for key, value in pref.iteritems():
+                        if not user_profile.get(key):
+                            continue
+                        user_data = user_profile.get(key) or 0
+                        try:
+                            user_data = float(user_data)
+                            score += value * user_data
+                        except ValueError as e:
+                            # TODO: when user profile is not number, we need another method to calculate score
```

Comment:
When would user profile not be a number? If user profile information isn't available don't we assign it 0 by default?